File: code/lt0838.java
// https://leetcode-cn.com/problems/push-dominoes/
// push dominoes
class Solution {
    public String pushDominoes(String dominoes) {
        int len = dominoes.length();
        char[] temp = dominoes.toCharArray();
        int p = -1;
        for(int i = 0; i < len; i++){
            char cur = temp[i];
            if(cur == '.')
                continue;
            char pre = p!=-1 ? temp[p] : 'L';
            
            if(cur == pre){
                for(int j = p + 1; j < i; j++)
                    temp[j] = cur;
            }
            
            if(pre == 'R' && cur == 'L'){
                int l = p + 1;
                int r = i - 1;
                while(l < r){
                    temp[l++] = 'R';
                    temp[r--] = 'L';
                }
            }
            
            p = i;
        }
        
        if(p!=-1 && temp[p] == 'R'){
            for(int i = p + 1; i < len; i++)
                temp[i] = 'R';
        }
        
        return String.valueOf(temp);
    }
}

File: code/lt0838.cpp
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    string pushDominoes(string dominoes) {
        
        int i = 0;
        while(dominoes[i]=='.')
            i++;
        int pre = i;

        if(dominoes[pre] == 'L')
        {
            for(int j = 0; j < i; j++)
                dominoes[j] = 'L';
        }
        
        int size = dominoes.size();

        for(i = i + 1; i < size ;i++)
        {
            char cur = dominoes[i]; 
            if(cur  =='.')
                continue;
            
            if(cur == dominoes[pre])
            {
                for(int j = pre + 1; j < i; j++)
                    dominoes[j] = cur;
            }
            else
            {
                int len = (i - pre -1)/2;
                if(cur == 'L')
                {
                    for(int j = 0; j < len; j++)
                    {
                        dominoes[pre +j + 1] = dominoes[pre];
                        dominoes[i - j - 1] = cur;
                    }
                }
            }
        
            pre = i;
        }
        
        if(dominoes[pre] =='R')
        {
           for(int i = pre; i < size; i++)
               dominoes[i] = 'R';
               
        }
        
        return dominoes;
    }
};